### ISSUE TITLE
[Docs] Lazy loading annotations

### ISSUE THREAD

**Issue by austinmw** [open-mmlab/mmdetection#9721]
### Branch

master branch  https://mmdetection.readthedocs.io/en/latest/

### 📚 The doc issue

Does MMDetection currently support lazy loading annotations when they don't all fit into memory (ex. loading a coco train_2017.json results in OOM)? If so, what's the easiest way to enable this?

**Comment by PeterVennerstrom** [comment]
The annotations could be saved as per image files containing a single image's annotations and read in during training.  The custom_dataset class could be adjusted not to pre-load the annotations and the [get_ann_info()](https://github.com/open-mmlab/mmdetection/blob/c14dd6c42efb63f662a63fe403198bac82f47aa6/mmdet/datasets/custom.py#L145) method adjusted to read them in from file by index. 

If masks aren't used they could be removed.  RLE encoding for masks may reduce memory.  COCO panoptic datasets store the mask data in png files which are read in as needed.

**Comment by austinmw** [comment]
So to make sure, creating a custom dataset does not require the `CustomDataset.load_annotations` method to return all annotations? I see that the `__init__` uses it to create `data_infos`, so I could remove that. I just want to make sure there's no other dependencies related to callbacks, evaluation, etc. that requires loading everything upfront.

**Comment by PeterVennerstrom** [comment]
I'd guess most of the dependancies are in terms of the images not the annotations.  The existing code stores image/annotations in per image dictionaries within the list assigned to `self.data_infos`.

`CustomDataset` inherits [torch Dataset](https://pytorch.org/docs/stable/data.html#torch.utils.data.Dataset) which requires `__getitem__` and `__len__`.  They have some nice [documentation](https://pytorch.org/tutorials/beginner/basics/data_tutorial.html).  Torch `IterableDataset` is also option to build your dataset, but I think the mmdet datasets are built on the map-style only.

The partitioning for distributed training is [here](https://github.com/open-mmlab/mmdetection/blob/master/mmdet/datasets/samplers/distributed_sampler.py).  It inherits from the torch `DistributedSampler`.

WebDataset requires a torch `IterableDataset`.

**Comment by hhaAndroid** [comment]
@austinmw The main branch is currently not supported, but the 3.x branch supports it

```python
test_dataset_cfg = copy.deepcopy(config.test_dataloader.dataset)
# lazy init. We only need the metainfo.
test_dataset_cfg['lazy_init'] = True

DATASETS.build(test_dataset_cfg)
```

**Comment by Looottch** [comment]
@hhaAndroid 
can you give a code link for lazy_init here
> @austinmw The main branch is currently not supported, but the 3.x branch supports it
> 
> ```python
> test_dataset_cfg = copy.deepcopy(config.test_dataloader.dataset)
> # lazy init. We only need the metainfo.
> test_dataset_cfg['lazy_init'] = True
> 
> DATASETS.build(test_dataset_cfg)
> ```

**Comment by Bing1002** [comment]
Any plan to integrate webdataset as another dataloader option? @hhaAndroid

**Comment by KyanChen** [comment]
Any plan to integrate webdataset as another dataloader option?
